Catches missing 'UNITS' attribute while loading CDF files (and adds some _known_units)

jgieseler opened the pull request:
## PR Description

`read_cdf()` from `sunpy.io.cdf` addresses the variable attribute `'UNITS'` without checking if it exists. But some CDF files (especially of older missions, e.g., STEREO/LET) don't follow the ISTP guide on CDF files, and don't provide `'UNITS'`. This PR fixes this problem by catching the resulting KeyError, assigning dimensionless units to the corresponding variables, and providing a SunpyUserWarning.

Fixes #5907

As a small addition, 5 variations of units for STEREO/LET CDF data files are added to the dictionary of `_known_units` used by `read_cdf()`. (Similar to #5853)

## Example

``` python
>>> import warnings
>>> from sunpy.net import Fido
>>> from sunpy.net import attrs as a
>>> from sunpy.io.cdf import read_cdf

>>> # omit Pandas' PerformanceWarning here because they originate somewhere else
>>> warnings.simplefilter(action='ignore', category=pd.errors.PerformanceWarning)

>>> trange = a.Time('2021/07/01', '2021/07/02')
>>> dataset = a.cdaweb.Dataset('STA_L1_LET')
>>> result = Fido.search(trange, dataset)
>>> downloaded_files = Fido.fetch(result)

>>> data = read_cdf(downloaded_files[0])
WARNING: SunpyUserWarning: No units provided for variable "CodeOK". Assigning dimensionless units. If you think this variable should not be without units, please raise an issue at https://github.com/sunpy/sunpy/issues [sunpy.io.cdf]
WARNING: SunpyUserWarning: No units provided for variable "DyThState". Assigning dimensionless units. If you think this variable should not be without units, please raise an issue at https://github.com/sunpy/sunpy/issues [sunpy.io.cdf]
WARNING: SunpyUserWarning: No units provided for variable "LeakConv". Assigning dimensionless units. If you think this variable should not be without units, please raise an issue at https://github.com/sunpy/sunpy/issues [sunpy.io.cdf]
WARNING: SunpyUserWarning: No units provided for variable "SWVersion". Assigning dimensionless units. If you think this variable should not be without units, please raise an issue at https://github.com/sunpy/sunpy/issues [sunpy.io.cdf]
>>> print(data)
[<sunpy.timeseries.timeseriesbase.GenericTimeSeries object at 0x7f45a75fc5b0>]
```

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
- [x] I have followed the guidelines in the [Contributing document](https://docs.sunpy.org/en/latest/dev_guide/contents/newcomers.html)
- [x] Changes follow the coding style of this project
- [x] Changes have been formatted and linted
- [x] Changes pass pytest style unit tests (and `pytest` passes).
- [x] Changes include any required corresponding changes to the documentation
- [x] Changes have been rebased and squashed to the minimal number (typically 1) of relevant commits
- [x] Changes have a descriptive commit message with a short title
- [x] I have added a `Fixes #XXXX -` or `Closes #XXXX -` comment to auto-close the issue that your PR addresses

jgieseler commented:
pre-commit.ci autofix

dstansby commented on sunpy/io/cdf.py:
```diff
@@ -71,18 +71,28 @@ def read_cdf(fname):
             data = cdf.varget(var_key)
             # Get units
-            unit_str = attrs['UNITS']
             try:
```
Instead of catching a `KeyError` here, I think it would be better to read with `if 'UNITS' in attrs:` instead.

jgieseler replied:
Yes, that's actually more elegant. Don't know why I took a more complicated way. 😬

dstansby commented on sunpy/io/cdf.py:
```diff
@@ -71,18 +71,28 @@ def read_cdf(fname):
+                    warn_user(f'No units provided for variable "{var_key}". '
                               'Assigning dimensionless units. '
-                              'If you think this unit should not be dimensionless, '
+                              'If you think this variable should not be without units, '
```
In the message I don't think we should tell users to raise an issue, as we don't want to be maintaining a list of all variables names out there that don't have a units field.

jgieseler replied:
Was thinking about that, too. Reducing it to `'No units provided for variable "{var_key}". Assigning dimensionless units.'`

dstansby reviewed (changes requested):
Looks good, apart from the way in which we detect the lack of units field.

jgieseler commented:
> Looks good, apart from the way in which we detect the lack of units field.

Introduced requested changes (and added 4 more `_known_units` to the zoo of STEREO variations...)
